### PR TITLE
fix(launchpad): correctly migrate projects settings config in plugins

### PR DESCRIPTION
**Pull request by lmiller1990**
- Closes https://cypress-io.atlassian.net/browse/UNIFY-1183

### User facing changelog

Correctly consider config values set when executing `pluginsFile` during migration. For example:

```js
module.exports = (on, config) => {
  config.integrationFolder = 'tests/e2e'
  return config
}
```

We will now correctly modify your new `cypress.config.js` to contain `specPattern: 'tests/e2e/**/*.cy.{js,jsx,ts,tsx}'`.

We simply execute the plugins and merge the returned value with `cypress.json`. We pass `on` as a no-op, since we don't actually want to execute any of the tasks the user might have defined during migration.

### Additional details

It's necessary to have a smooth migration.

### PR Tasks

- [x] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [ ] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

File: packages/data-context/src/sources/migration/codegen.ts [hunk 1]
```diff
@@ -55,7 +36,7 @@ export interface CreateConfigOptions {
   hasTypescript: boolean
 }
 
-export async function createConfigString (cfg: OldCypressConfig, options: CreateConfigOptions) {
+export async function createConfigString (cfg: Partial<Cypress.Config>, options: CreateConfigOptions) {
```
Comment (elevatebart) [inline packages/data-context/src/sources/migration/codegen.ts:39]: Why did you change this type?

`integrationFolder` & `componentFolder` should not work with `Cypress.Config`...

Although maybe they do and that is a separate issue

File: packages/graphql/src/schemaTypes/objectTypes/gql-Migration.ts [hunk 2]
```diff
@@ -20,15 +20,15 @@ export const MigrationStep = objectType({
...
       },
     })
 
     t.nonNull.boolean('isCompleted', {
       description: 'Has the current step been completed',
       resolve: (source, args, ctx) => {
-        const indexOfObservedStep = ctx.migration.filteredSteps.indexOf(source.name)
-        const indexOfCurrentStep = ctx.migration.filteredSteps.indexOf(ctx.migration.step)
+        const indexOfObservedStep = ctx.coreData.migration.filteredSteps.indexOf(source.name)
+        const indexOfCurrentStep = ctx.coreData.migration.filteredSteps.indexOf(ctx.coreData.migration.step)
```
Comment (elevatebart) [inline packages/graphql/src/schemaTypes/objectTypes/gql-Migration.ts:31]: Why change that?

We remove a layer of abstraction that IMHO has a very real use in the graphql package.

**Review by elevatebart — CHANGES_REQUESTED** [review]
A couple of change requests. 1 big question:

After digging deep in your PR, there are a lot of changes in this PR that do not pertain to reading the pluginsFile in the mgration process:
- Variable name changes
- Type refactoring
- gql-Migration changes

Were they necessary? helpful? should they be in a separate PR?

File: packages/data-context/src/sources/migration/codegen.ts (hunk 1 above)
Comment (lmiller1990) [reply packages/data-context/src/sources/migration/codegen.ts:39]: The type `OldCypressConfig` is wrong - we receiving the legacy config here as this arg, fully including `integrationFolder`, even if it's not included in the type. I'm not sure why this type was created in the first place?

`integrationFolder` etc get removed later in `reduceConfig` - they aren't included in the created `cypress.config.js`.

File: packages/graphql/src/schemaTypes/objectTypes/gql-Migration.ts (hunk 2 above)
Comment (lmiller1990) [reply packages/graphql/src/schemaTypes/objectTypes/gql-Migration.ts:31]: I just moved the data from getting stored in a Source to CoreData. Isn't the idea of a source to derive data from coreData, instead of having multiple places to save things?

The problem I encountered is we need to make sure we fully derive the legacyConfig include the modification from `cypress/plugins/index.js`. This has to happen before we call `initializeFlags`. 

I also found it quite confusing we have data in multiple places - `coreData`, `lifecycleManager` and Sources. We could  have `steps` in migration and derive a subset of `filteredSteps` based on the flags, if that's what you are suggesting here?

**Comment by lmiller1990** [comment]
> A couple of change requests. 1 big question:
>
> After digging deep in your PR, there are a lot of changes in this PR that do not pertain to reading the pluginsFile in the mgration process:

>    Variable name changes
>    Type refactoring
>    gql-Migration changes

> Were they necessary? helpful? should they be in a separate PR?

`OldCypressConfig` was incorrect - the object we get there has way more properties (all legacy ones). `OldCypressConfig` only had a subset, which was just incorrect. I changed it to use the actual type it gets, which is just a Partial CypressConfig. It reveals some bugs - eg `integrationFolder` cannot be false.

I don't think I change anything in the GraphQL - I realized that when we call `initializeFlags` we need to have the legacy config available (including execute plugins) which we didn't already have. I also think we should move towards the architecture of data in coreData, DataSources simply **derive** data (and don't set any new values) and actions is where all the mutation occurs. It's quite confusing how Sources and Actions kind of do the same thing right now.

Strictly speaking I guess I could have potentially done this with less code changes, but do you think the changes here introduce significant risk/make the PR too big?

File: packages/data-context/src/actions/MigrationActions.ts [hunk 3]
```diff
@@ -7,12 +9,141 @@ import {
...
+  isDefaultTestFiles,
+  getComponentTestFilesGlobs,
+  getComponentFolder,
+  getIntegrationTestFilesGlobs,
+  getSpecPattern,
+} from '../sources/migration'
+
+export function processConfigViaLegacyPlugins (projectRoot: string, legacyConfig: Partial<Cypress.Config>): Promise<Partial<Cypress.Config>> {
+  return new Promise(async (resolve, reject) => {
+    const pluginFile = legacyConfig.pluginsFile ?? await tryGetDefaultLegacyPluginsFile(projectRoot)
```
Comment (ryanthemanuel) [inline packages/data-context/src/actions/MigrationActions.ts:28]: Seeing async functions inside of a promise constructor tends to make me nervous. I believe we would be swallowing any errors that might be thrown from tryGetDefaultLegacyPluginsFile. Can we do something like this instead?

```suggestion
export async function processConfigViaLegacyPlugins (projectRoot: string, legacyConfig: Partial<Cypress.Config>): Promise<Partial<Cypress.Config>> {
  const pluginFile = legacyConfig.pluginsFile ?? await tryGetDefaultLegacyPluginsFile(projectRoot)
  return new Promise((resolve, reject) => {
```

File: packages/data-context/src/sources/migration/codegen.ts [hunk 4]
```diff
@@ -21,25 +21,6 @@ type ConfigOptions = {
...
-  viewportWidth?: number
-  baseUrl?: string
-  retries?: number
-
-  component?: Omit<OldCypressConfig, 'component' | 'e2e'>
-  e2e?: Omit<OldCypressConfig, 'component' | 'e2e'>
-  pluginsFile?: string | false
-  supportFile?: string | false
-  componentFolder?: string | false
-  integrationFolder?: string | false
```
Comment (lmiller1990) [inline packages/data-context/src/sources/migration/codegen.ts:38]: This cannot be false

File: packages/server/lib/plugins/child/run_require_async_child.js [hunk 5]
```diff
@@ -72,16 +72,59 @@ function run (ipc, configFile, projectRoot) {
...
+      ipc.send('loadLegacyPlugins:reply', { config: legacyConfig })
+
+      return
+    }
+
+    // we do not want to execute any tasks - the purpose
+    // of this is to get any modified config returned
+    // by plugins.
+    const noop = () => {}
+    const legacyPluginsConfig = await legacyPlugins(noop, legacyConfig)
```
Comment (lmiller1990) [inline packages/server/lib/plugins/child/run_require_async_child.js:99]: Should error here?

Comment (lmiller1990) [reply packages/server/lib/plugins/child/run_require_async_child.js:99]:
```
reply(setupTestingType:error)
```

**Comment by lmiller1990** [comment]
@ryanthemanuel done.
@elevatebart we synced and made your changes.

Also added test to make sure we handle errors that might be thrown during migration when we execute the pluginsFile (highly unlikely, unless they have an already broken pluginsFile).

image

File: packages/data-context/src/actions/MigrationActions.ts (hunk 3 above)
Comment (lmiller1990) [reply packages/data-context/src/actions/MigrationActions.ts]: done!

**Comment by lmiller1990** [comment]
@elevatebart @ryanthemanuel I updated to have more tests and made the [ipc between legacy plugins type safe](https://github.com/cypress-io/cypress/pull/20509/files#diff-0217118b16e48b5b3cc2e641371ac333285c32a1ebc9bbb7d773a3c1d48305deR30-R99). I also did a back merge and resolved conflicts. Please take another look, thanks!

File: packages/data-context/src/sources/MigrationDataSource.ts [hunk 6]
```diff
@@ -40,81 +43,25 @@ export interface MigrationFile {
...
-  hasCustomIntegrationFolder: boolean = flags.hasCustomIntegrationFolder
-  hasCustomIntegrationTestFiles: boolean = flags.hasCustomIntegrationTestFiles
-
-  hasCustomComponentFolder: boolean = flags.hasCustomComponentFolder
-  hasCustomComponentTestFiles: boolean = flags.hasCustomComponentTestFiles
-
-  hasCustomSupportFile: boolean = flags.hasCustomSupportFile
-  hasComponentTesting: boolean = flags.hasComponentTesting
-  hasE2ESpec: boolean = flags.hasE2ESpec
-  hasPluginsFile: boolean = flags.hasPluginsFile
```
Comment (tgriesser) [inline packages/data-context/src/sources/MigrationDataSource.ts:72]: Thank you for making this class stateless!

**Review by tgriesser — CHANGES_REQUESTED** [review]
Looks great - was just going through and auditing issues that might be related to lifecycle management and saw that I completely missed being added for review here. Glad this stuff is getting cleaned up a bit & we're moving a lot the state out of the `MigrationDataSource`. 

1 minor suggestion, but otherwise 👍

File: packages/data-context/src/data/ProjectLifecycleManager.ts [hunk 7]
```diff
@@ -272,6 +275,22 @@ export class ProjectLifecycleManager {
 
     const { needsCypressJsonMigration } = this.refreshMetaState()
 
+    const cfgPath = path.join(projectRoot, this.legacyConfigFile)
```
Comment (estrada9166) [inline packages/data-context/src/data/ProjectLifecycleManager.ts:278]:
```suggestion
    const legacyCfgPath = this.legacyJsonPath
```

File: packages/data-context/src/actions/MigrationActions.ts [hunk 8]
```diff
@@ -152,4 +334,19 @@ export class MigrationActions {
...
+      coreData.migration.flags = {
+        hasCustomIntegrationFolder: false,
+        hasCustomIntegrationTestFiles: false,
+        hasCustomComponentFolder: false,
+        hasCustomComponentTestFiles: false,
+        hasCustomSupportFile: false,
+        hasComponentTesting: true,
+        hasE2ESpec: true,
+        hasPluginsFile: true,
+      }
```
Comment (estrada9166) [inline packages/data-context/src/actions/MigrationActions.ts:349]: Should we set the default values as a constant in `packages/data-context/src/data/coreDataShape.ts` and use that one in both places? so we do not miss anything?

File: packages/server/lib/plugins/child/run_require_async_child.js [hunk 9]
```diff
@@ -72,16 +72,65 @@ function run (ipc, configFile, projectRoot) {
...
+      const noop = () => {}
+      const legacyPluginsConfig = await legacyPlugins(noop, legacyConfig)
+
+      // match merging strategy from 9.x
+      const mergedLegacyConfig = {
+        ...legacyConfig,
+        ...legacyPluginsConfig,
+      }
+
+      if (legacyConfig.e2e || legacyPluginsConfig.e2e) {
```
Comment (estrada9166) [inline packages/server/lib/plugins/child/run_require_async_child.js:108]: If the user's plugin do not return `config` which I think can happen based on [this](https://docs.cypress.io/guides/tooling/plugins-guide#Using-a-plugin) - we are throwing an error, we should take into account when this happens and `legacyPluginsConfig` is `undefined`

image

File: packages/data-context/src/actions/MigrationActions.ts [hunk 10]
```diff
@@ -1,22 +1,195 @@
...
+}
 
 export class MigrationActions {
   constructor (private ctx: DataContext) { }
 
+  async initialize (config: LegacyCypressConfigJson) {
+    const legacyConfigForMigration = await this.setLegacyConfigForMigration(config)
+
+    // for testing mainly, we want to ensure the flags are reset each test
+    this.resetFlags()
```
Comment (estrada9166) [inline packages/data-context/src/actions/MigrationActions.ts:109]: Should we add this to `clearCurrentProject`?

**Review by estrada9166 — CHANGES_REQUESTED** [review]

File: packages/data-context/src/actions/MigrationActions.ts (hunk 10 above)
Comment (lmiller1990) [reply packages/data-context/src/actions/MigrationActions.ts:109]: Hm - I don't think so. `clearCurrentProject` is involved with getting rid of everything to do with the current project, `resetFlags` is called by `MigrationActions#initialize` which is called by `setCurrentProject`. This seems like a reasonable place to reset the state.

File: packages/server/lib/plugins/child/run_require_async_child.js (hunk 9 above)
Comment (lmiller1990) [reply packages/server/lib/plugins/child/run_require_async_child.js:116]: added test and fixed https://github.com/cypress-io/cypress/pull/20509/commits/5f11b6ba8c637d67b0efc608d0973f95234e47d3

File: packages/data-context/src/data/ProjectLifecycleManager.ts (hunk 7 above)
Comment (lmiller1990) [reply packages/data-context/src/data/ProjectLifecycleManager.ts]: https://github.com/cypress-io/cypress/pull/20509/commits/a5c5b55ceaa383be48892d0b3dab0cf148b96b47

File: packages/data-context/src/actions/MigrationActions.ts (hunk 8 above)
Comment (lmiller1990) [reply packages/data-context/src/actions/MigrationActions.ts]: sure, we can just reset using the defaults: https://github.com/cypress-io/cypress/pull/20509/commits/ec7f9ea7d9c23051efca87d5ed4efbdce7e545bb

File: packages/data-context/src/actions/MigrationActions.ts [hunk 11]
```diff
@@ -1,22 +1,195 @@
...
+  isDefaultTestFiles,
+  getComponentTestFilesGlobs,
+  getComponentFolder,
+  getIntegrationTestFilesGlobs,
+  getSpecPattern,
+} from '../sources/migration'
+import EventEmitter from 'events'
+import type { CypressError } from '@packages/errors'
+
+class LegacyPluginsIpc extends EventEmitter {
```
[the remark this answers is not in the document]
Comment (lmiller1990) [reply packages/data-context/src/actions/MigrationActions.ts]: Ok I did it here https://github.com/cypress-io/cypress/pull/20509/commits/fa8327f235e86f12d2bbde73b80860fdf4f63463

File: packages/server/lib/plugins/child/run_require_async_child.js (hunk 5 above)
Comment (lmiller1990) [reply packages/server/lib/plugins/child/run_require_async_child.js]: fixed

**Review by elevatebart — APPROVED** [review]

**Comment by lmiller1990** [comment]
@ryanthemanuel @estrada9166 @tgriesser I fixed everything up, please re-review.

**Review by ryanthemanuel — APPROVED** [review]

**Review by tgriesser — APPROVED** [review]
Nice work!

**Review by estrada9166 — APPROVED** [review]
